Source/Dafny/DafnyAst.cs

```diff
@@ -12544,6 +12569,38 @@ public AutoGeneratedExpression(IToken tok, Expression e)
     }
   }
 
+  /// <summary>
+  /// When an actual parameter is omitted for a formal with a default value, the positional resolved
+  /// version of the actual parameter will have a DefaultValueExpression value. This has three
+  /// advantages:
+  /// * It allows the entire module to be resolved before any substitutions take place.
+  /// * It gives a good place to check for default-value expressions that would give rise to an
+  ///   infinite expansion.
+  /// * It preserves the pre-substitution form, which gives compilers a chance to avoid re-evaluation
+  ///   of actual parameters used in other default-valued expressions.
+  /// </summary>
+  public class DefaultValueExpression : ConcreteSyntaxExpression {
```

Comment:
I'll buy that this class is useful, but does it have to be a subtype of `Expression` itself? That feels wrong to me, and just invites bugs where this type of "expression" can live elsewhere in the AST where it doesn't belong.

Reply:
Good thought. Indeed, in working on this PR, I first missed some places where these new expressions crept in.

I tried to change it now. I find there are several places that take lists of `Expression`-arguments that then would need to become more complicated lists. So, I'm still okay with it being a `ConcreteSyntaxExpression` subclass.

Reply:
Perhaps this is just worth a TODO acknowledging that the modelling is slightly off, just so other readers of the code can keep it in mind?